sandbox/libs/authn/src/main/java/org/opensearch/authn/realm/InternalRealm.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.authn.realm;
+
+import org.apache.shiro.authc.AuthenticationException;
+import org.apache.shiro.authc.AuthenticationInfo;
+import org.apache.shiro.authc.AuthenticationToken;
+import org.apache.shiro.authc.CredentialsException;
+import org.apache.shiro.authc.IncorrectCredentialsException;
+import org.apache.shiro.authc.SimpleAuthenticationInfo;
+import org.apache.shiro.authc.UnknownAccountException;
+import org.apache.shiro.authc.UsernamePasswordToken;
+import org.apache.shiro.realm.AuthenticatingRealm;
+import org.opensearch.authn.InternalSubject;
+
+import java.util.Objects;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * Internal Realm is a custom realm using the internal OpenSearch IdP
+ *
+ * @opensearch.experimental
+ */
+public class InternalRealm extends AuthenticatingRealm {
+    private static final String DEFAULT_REALM_NAME = "internal";
+
+    public static final InternalRealm INSTANCE = new InternalRealm.Builder(DEFAULT_REALM_NAME).build();
+
+    private final String realmName;
+
+    private InternalRealm(String realmName) {
+        super(new BCryptPasswordMatcher());
+        this.realmName = realmName;
+    }
+
+    // TODO Switch this to private after debugging
+    public ConcurrentMap<String, InternalSubject> internalSubjects;
+
+    public static final class Builder {
+        private final String name;
+
+        public Builder(String name) {
+            this.name = Objects.requireNonNull(name);
+        }
+
+        public InternalRealm build() {
+            return new InternalRealm(name);
+        }
+    }
+
+    public void initializeInternalSubjectsStore(String pathToInternalUsersYaml) {
```

Comment:
With the builder this method can be called during `build()`, want to just inline it there?

Reply:
Done